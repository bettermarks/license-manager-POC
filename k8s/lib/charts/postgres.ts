import { Chart, ChartProps } from "cdk8s";
import {
  ConfigMap,
  ImagePullPolicy,
  ISecret,
  Secret,
  ServiceType,
} from "cdk8s-plus-24";
import { Construct } from "constructs";
import { IntOrString, KubeDeployment, KubeService } from "../../imports/k8s";

interface PostgresChartProps extends ChartProps {
  image: string;
  name: string;
}

export class PostgresChart extends Chart {
  /**
   * Secret with Postgres credentials
   *
   * keys:
   * - DATABASE_HOST
   * - DATABASE_PORT
   * - DATABASE_USER
   * - DATABASE_PASSWORD
   * - DATABASE_NAME
   * - POSTGRES_PASSWORD
   */
  readonly secret: ISecret;

  constructor(scope: Construct, id: string, props: PostgresChartProps) {
    super(scope, id, props);
    const { name, namespace } = props;
    const pgSecret = new Secret(this, "secret", {
      metadata: {
        name: name,
        namespace,
      },
      stringData: {
        DATABASE_PORT: "5432",
        DATABASE_USER: "postgres",
        DATABASE_PASSWORD: "postgres",
        DATABASE_NAME: "postgres",
        POSTGRES_PASSWORD: "postgres",
        POSTGRES_USER: "postgres",
        POSTGRES_DB: "postgres",
      },
    });

    const configMap = new ConfigMap(this, "initdb", {
      metadata: {
        name: `${name}-init-db-script`,
        namespace,
      },
      data: {}, // TODO: any init command?
    });

    new KubeDeployment(this, "pg-deployment", {
      metadata: {
        name,
        namespace,
        labels: {
          app: name,
        },
      },
      spec: {
        selector: {
          matchLabels: {
            app: name,
          },
        },
        replicas: 1,
        template: {
          metadata: {
            labels: {
              app: name,
            },
          },
          spec: {
            containers: [
              {
                name,
                image: props.image,
                imagePullPolicy: ImagePullPolicy.IF_NOT_PRESENT,
                ports: [{ containerPort: 5432 }],
                envFrom: [
                  {
                    secretRef: { name: pgSecret.name },
                  },
                ],
                readinessProbe: {
                  tcpSocket: {
                    port: IntOrString.fromNumber(5432),
                  },
                  initialDelaySeconds: 10,
                },
                livenessProbe: {
                  tcpSocket: {
                    port: IntOrString.fromNumber(5432),
                  },
                  initialDelaySeconds: 10,
                },
                volumeMounts: [
                  {
                    name: `${configMap.name}-volume`,
                    mountPath: "/docker-entrypoint-initdb.d/",
                  },
                ],
              },
            ],
            volumes: [
              {
                name: `${configMap.name}-volume`,
                configMap: {
                  name: configMap.name,
                },
              },
            ],
          },
        },
      },
    });

    const pgService = new KubeService(this, id, {
      metadata: {
        name,
        namespace,
      },
      spec: {
        type: ServiceType.CLUSTER_IP,
        ports: [
          {
            port: 5432,
          },
        ],
        selector: {
          app: name,
        },
      },
    });

    pgSecret.addStringData("DATABASE_HOST", pgService.name);
    // pgSecret.addStringData("DATABASE_PORT", "5432");
    this.secret = pgSecret;
  }
}
