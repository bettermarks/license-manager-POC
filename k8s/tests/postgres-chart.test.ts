import { Testing } from "cdk8s";
import { PostgresChart } from "../lib/postgres/postgres-chart";

describe("postgres-chart", () => {
  test("should match with snapshot", () => {
    const app = Testing.app();
    const chart = new PostgresChart(app, "test-chart", { image: "postgres:14" });
    const results = Testing.synth(chart);
    expect(results).toMatchSnapshot();
  });
});
