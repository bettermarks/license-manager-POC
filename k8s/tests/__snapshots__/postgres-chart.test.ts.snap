// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`postgres-chart should match with snapshot 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "immutable": false,
    "kind": "Secret",
    "metadata": Object {
      "name": "licensing-db",
      "namespace": "dummy-namespace",
    },
    "stringData": Object {
      "DATABASE_HOST": "licensing-db",
      "DATABASE_NAME": "postgres",
      "DATABASE_PASSWORD": "postgres",
      "DATABASE_PORT": "5432",
      "DATABASE_USER": "postgres",
      "POSTGRES_DB": "postgres",
      "POSTGRES_PASSWORD": "postgres",
      "POSTGRES_USER": "postgres",
    },
  },
  Object {
    "apiVersion": "v1",
    "immutable": false,
    "kind": "ConfigMap",
    "metadata": Object {
      "name": "licensing-db-init-db-script",
      "namespace": "dummy-namespace",
    },
  },
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "labels": Object {
        "app": "licensing-db",
      },
      "name": "licensing-db",
      "namespace": "dummy-namespace",
    },
    "spec": Object {
      "replicas": 1,
      "selector": Object {
        "matchLabels": Object {
          "app": "licensing-db",
        },
      },
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "app": "licensing-db",
          },
        },
        "spec": Object {
          "containers": Array [
            Object {
              "envFrom": Array [
                Object {
                  "secretRef": Object {
                    "name": "licensing-db",
                  },
                },
              ],
              "image": "postgres:14",
              "imagePullPolicy": "IfNotPresent",
              "livenessProbe": Object {
                "initialDelaySeconds": 10,
                "tcpSocket": Object {
                  "port": 5432,
                },
              },
              "name": "licensing-db",
              "ports": Array [
                Object {
                  "containerPort": 5432,
                },
              ],
              "readinessProbe": Object {
                "initialDelaySeconds": 10,
                "tcpSocket": Object {
                  "port": 5432,
                },
              },
              "volumeMounts": Array [
                Object {
                  "mountPath": "/docker-entrypoint-initdb.d/",
                  "name": "licensing-db-init-db-script-volume",
                },
              ],
            },
          ],
          "volumes": Array [
            Object {
              "configMap": Object {
                "name": "licensing-db-init-db-script",
              },
              "name": "licensing-db-init-db-script-volume",
            },
          ],
        },
      },
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": Object {
      "name": "licensing-db",
      "namespace": "dummy-namespace",
    },
    "spec": Object {
      "ports": Array [
        Object {
          "port": 5432,
        },
      ],
      "selector": Object {
        "app": "licensing-db",
      },
      "type": "ClusterIP",
    },
  },
]
`;
